staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go

```diff
@@ -425,9 +425,10 @@ func GetPodRunningTimeoutFlag(cmd *cobra.Command) (time.Duration, error) {
 type FeatureGate string
 
 const (
-	ApplySet              FeatureGate = "KUBECTL_APPLYSET"
-	CmdPluginAsSubcommand FeatureGate = "KUBECTL_ENABLE_CMD_SHADOW"
-	InteractiveDelete     FeatureGate = "KUBECTL_INTERACTIVE_DELETE"
+	ApplySet                FeatureGate = "KUBECTL_APPLYSET"
+	CmdPluginAsSubcommand   FeatureGate = "KUBECTL_ENABLE_CMD_SHADOW"
+	InteractiveDelete       FeatureGate = "KUBECTL_INTERACTIVE_DELETE"
+	RemoteCommandWebsockets FeatureGate = "KUBECTL_REMOTE_COMMAND_WEBSOCKETS"
```

Comment:
asking because I really don't know, do we have e2e test for these?

Reply:
I think we can add an e2e test just copying this test and setting the corresponding environment variable

https://github.com/kubernetes/kubernetes/blob/df5d84ae811bc119e715ee39dafea5067dd32f77/test/e2e/kubectl/kubectl.go#L593-L651

The e2e test has to work always, independetly of the apiserver

Reply:
@seans3 this should make it

```diff
diff --git a/test/e2e/kubectl/kubectl.go b/test/e2e/kubectl/kubectl.go
index 55639552c86..e33e6f02171 100644
--- a/test/e2e/kubectl/kubectl.go
+++ b/test/e2e/kubectl/kubectl.go
@@ -42,6 +42,8 @@ import (
 
        "sigs.k8s.io/yaml"
 
+       utilkubectl "kubectl/pkg/cmd/util"
+
        v1 "k8s.io/api/core/v1"
        rbacv1 "k8s.io/api/rbac/v1"
        apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
@@ -801,6 +803,65 @@ metadata:
                        framework.ExpectNoError(c.CoreV1().Pods(ns).Delete(ctx, "run-test-3", metav1.DeleteOptions{}))
                })
 
+               ginkgo.It("should support inline execution and attach using Websocket", func(ctx context.Context) {
+                       os.Setenv(utilkubectl.RemoteCommandWebsockets, "true")
+                       waitForStdinContent := func(pod, content string) string {
+                               var logOutput string
+                               err := wait.Poll(10*time.Second, 5*time.Minute, func() (bool, error) {
+                                       logOutput = e2ekubectl.RunKubectlOrDie(ns, "logs", pod)
+                                       return strings.Contains(logOutput, content), nil
+                               })
+
+                               framework.ExpectNoError(err, "waiting for '%v' output", content)
+                               return logOutput
+                       }
+
+                       ginkgo.By("executing a command with run and attach with stdin")
+                       // We wait for a non-empty line so we know kubectl has attached
+                       e2ekubectl.NewKubectlCommand(ns, "run", "run-test", "--image="+busyboxImage, "--restart=OnFailure", podRunningTimeoutArg, "--attach=true", "--stdin", "--", "sh", "-c", "echo -n read: && cat && echo 'stdin closed'").
+                               WithStdinData("value\nabcd1234").
+                               ExecOrDie(ns)
+
+                       runOutput := waitForStdinContent("run-test", "stdin closed")
+                       gomega.Expect(runOutput).To(gomega.ContainSubstring("read:value"))
+                       gomega.Expect(runOutput).To(gomega.ContainSubstring("abcd1234"))
+                       gomega.Expect(runOutput).To(gomega.ContainSubstring("stdin closed"))
+
+                       framework.ExpectNoError(c.CoreV1().Pods(ns).Delete(ctx, "run-test", metav1.DeleteOptions{}))
+
+                       ginkgo.By("executing a command with run and attach without stdin")
+                       // There is a race on this scenario described in #73099
+                       // It fails if we are not able to attach before the container prints
+                       // "stdin closed", but hasn't exited yet.
+                       // We wait 10 seconds before printing to give time to kubectl to attach
+                       // to the container, this does not solve the race though.
+                       e2ekubectl.NewKubectlCommand(ns, "run", "run-test-2", "--image="+busyboxImage, "--restart=OnFailure", podRunningTimeoutArg, "--attach=true", "--leave-stdin-open=true", "--", "sh", "-c", "cat && echo 'stdin closed'").
+                               WithStdinData("abcd1234").
+                               ExecOrDie(ns)
+
+                       runOutput = waitForStdinContent("run-test-2", "stdin closed")
+                       gomega.Expect(runOutput).ToNot(gomega.ContainSubstring("abcd1234"))
+                       gomega.Expect(runOutput).To(gomega.ContainSubstring("stdin closed"))
+
+                       framework.ExpectNoError(c.CoreV1().Pods(ns).Delete(ctx, "run-test-2", metav1.DeleteOptions{}))
+
+                       ginkgo.By("executing a command with run and attach with stdin with open stdin should remain running")
+                       e2ekubectl.NewKubectlCommand(ns, "run", "run-test-3", "--image="+busyboxImage, "--restart=OnFailure", podRunningTimeoutArg, "--attach=true", "--leave-stdin-open=true", "--stdin", "--", "sh", "-c", "cat && echo 'stdin closed'").
+                               WithStdinData("abcd1234\n").
+                               ExecOrDie(ns)
+
+                       runOutput = waitForStdinContent("run-test-3", "abcd1234")
+                       gomega.Expect(runOutput).To(gomega.ContainSubstring("abcd1234"))
+                       gomega.Expect(runOutput).ToNot(gomega.ContainSubstring("stdin closed"))
+
+                       g := func(pods []*v1.Pod) sort.Interface { return sort.Reverse(controller.ActivePods(pods)) }
+                       runTestPod, _, err := polymorphichelpers.GetFirstPod(f.ClientSet.CoreV1(), ns, "run=run-test-3", 1*time.Minute, g)
+                       framework.ExpectNoError(err)
+                       framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(ctx, c, runTestPod.Name, ns, time.Minute))
+
+                       framework.ExpectNoError(c.CoreV1().Pods(ns).Delete(ctx, "run-test-3", metav1.DeleteOptions{}))
+               })
+
                ginkgo.It("should contain last line of the log", func(ctx context.Context) {
                        podName := "run-log-test"
```

Reply:
just checking carefully, it needs to append `AppendEnv([]string{utilkubectl.RemoteCommandWebsockets, "true"}).` to all the kubectl commands instead of setting the global env var

Reply:
Added e2e tests which append `AppendEnv([]string{string(utilkubectl.RemoteCommandWebsockets), "true"})` to the three new tests. Please let me know what you think.

Reply:
you can resolve this